[PyTorch] Use 128-bit vectors for ARM64

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138275
* #138005
* #137918
* #137917
* #137916
* #137915
* #137914
* #137913
* #137912
* #137911
* #137661
* __->__ #137426
* #138744
* #138716
* #138655
* #138542
* #138486

The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

swolchok committed:
[PyTorch] Use 128-bit vectors for ARM64

The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -54,7 +54,7 @@ if(NOT BUILD_LITE_INTERPRETER)
 EXCLUDE(ATen_CORE_SRCS "${ATen_CORE_SRCS}" ${ATen_CORE_TEST_SRCS})
 
-file(GLOB base_h "*.h" "detail/*.h" "cpu/*.h" "cpu/vec/vec512/*.h" "cpu/vec/vec256/*.h" "cpu/vec/vec256/vsx/*.h" "cpu/vec/vec256/zarch/*.h" "cpu/vec/sve/*.h" "cpu/vec/*.h" "quantized/*.h" "functorch/*.h")
+file(GLOB base_h "*.h" "detail/*.h" "cpu/*.h" "cpu/vec/vec512/*.h" "cpu/vec/vec128/*.h" "cpu/vec/vec256/*.h" "cpu/vec/vec256/vsx/*.h" "cpu/vec/vec256/zarch/*.h" "cpu/vec/sve/*.h" "cpu/vec/*.h" "quantized/*.h" "functorch/*.h")
```
the headers in vec128 don't seem to be installing in my local testing and I don't know why; I thought this change would do it

swolchok replied:
found it, setup.py needs to be updated as well

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
another attempted linux aarch64 fix on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok commented:
CI is 100% green, please review

jgong5 commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -183,6 +163,9 @@ struct Vectorized {
            typename = std::enable_if_t<(sizeof...(Args) == size())>>
   Vectorized(Args... vals) : values{vals...}{
   }
+  Vectorized(T(&arr)[kSize]) {
```
Is const ref better here?

jgong5 commented on aten/src/ATen/cpu/vec/vec512/vec512_float.h:
```diff
@@ -40,6 +40,9 @@ template <> class Vectorized<float> {
     values = _mm512_setr_ps(val1, val2, val3, val4, val5, val6, val7, val8,
                             val9, val10, val11, val12, val13, val14, val15, val16);
   }
+  Vectorized(float (&arr)[16])
```
Do you want to add this new constructor to other data types than float too?

jgong5 commented on aten/src/ATen/cpu/vec/vec256/vec256_convert.h:
```diff
@@ -297,20 +297,35 @@ struct VecConvert<
 template <>
-struct VecConvert<float, 1, BFloat16, 1> {
-  static inline VectorizedN<float, 1> apply(
+struct VecConvert<float, 2, BFloat16, 1> {
```
You may want to keep the `struct VecConvert<float, 1, BFloat16, 1>` too since there are cases where we still want to use half of the lane for `BFloat16` when the call range is not large enough.

kimishpatel commented:
I will try to review this (before EOW) but if you want to be unblocked sooner, feel free to rely on another reviewer.

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

maajidkhann commented:
@pytorchbot label "module: arm"

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok commented:
tests are green again. will (be forced to, so don't worry about it) fix mergeability before landing, it's just because I landed the bottom diff of the stack to github and re-exported from internal repo, which doesn't have the diff landed yet.

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
math is hard on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
compiling code before you send it out for review is hard on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok commented:
can't figure out why we're decreasing accuracy for test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_group_norm_cpu_float16

swolchok commented:
inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCPU::test_comprehensive_prod_cpu_bool failure looks like we've confused inductor. The code snippet is: 

```
                auto tmp0 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(x1 + (25LL*x0)));
                auto tmp3 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(5LL + x1 + (25LL*x0)));
                auto tmp7 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(10LL + x1 + (25LL*x0)));
                auto tmp11 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(15LL + x1 + (25LL*x0)));
                auto tmp15 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(20LL + x1 + (25LL*x0)));
                auto tmp1 = tmp0.to<int64_t,2>();
                auto tmp2 = at::vec::VecMask<int64_t,2>::from<int64_t,2>(tmp1);
                auto tmp4 = tmp3.to<int64_t,2>();
                auto tmp5 = at::vec::VecMask<int64_t,2>::from<int64_t,2>(tmp4);
                auto tmp6 = tmp2 * tmp5;
```

and the `tmp6` multiplication doesn't build because `VecMask<T,N>` (of course) only implicitly converts to `Vectorized<T>` if   N=1. Presumably, it previously worked because of some sort of shenanigans where the float Vectorized instances were only half-full (single vector register), and therefore could be confused into single int64_t vectors (two vector registers).

swolchok commented:
seems like inductor might need the full slate of Vectorized operations on VectorizedN?

swolchok commented:
by contrast, here's the old, working code snippet:
```
            auto tmp0 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(x0));
            auto tmp2 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(25LL + x0));
            auto tmp5 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(50LL + x0));
            auto tmp8 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(75LL + x0));
            auto tmp11 = at::vec::VecMask<float,1>::from(in_ptr0 + static_cast<int64_t>(100LL + x0));
            auto tmp1 = tmp0.to<int64_t,2>();
            auto tmp3 = tmp2.to<int64_t,2>();
            auto tmp4 = tmp1 * tmp3;
```

so the problem is that for some reason we're converting from VectorizedN back to VecMask, I think?

swolchok commented:
> so the problem is that for some reason we're converting from VectorizedN back to VecMask, I think?

grepping torch/_inductor/cpp.py, it looks like before, we were converting from bool (represented as float?!) to int64. now, we're converting from bool (represented as float?!) to int64 and then back to bool. In other words, both of the conversions look like they have to be coming from CppVecKernel.get_to_dtype_expr.

swolchok commented:
@jgong5 / @leslie-fang-intel , is the right fix for the test_comprehensive_prod_cpu_bool failure to just define `operator*` on VecMask? I'm unsure because Inductor seems confused about VecMasks in general.

For example, the test_comprehensive_constant_pad_nd_cpu_bool failure is different: it's trying to apply VecMask::cast to a Vectorized<bool>. Here's the broken code snippet:

```
                    auto tmp23 = [&]
                    {
                        auto tmp24 = static_cast<bool>(false);
                        return tmp24;
                    }
                    ;
                    auto tmp30 =
                    [&]
                    {
                        if (tmp22.all_zero())
                        {
                            return at::vec::VecMask<float,1>::from(static_cast<bool>(true));
                        }
                        else
                        {
                            auto tmp25 = at::vec::Vectorized<bool>(tmp23());
                            auto tmp26 = at::vec::VecMask<float,1>::from(static_cast<bool>(true));
                            auto tmp27 = tmp22.template cast<float,1>();
                            auto tmp28 = tmp25.template cast<float,1>();
                            auto tmp29 = tmp26.template cast<float,1>();
                            return decltype(tmp28)::blendv(tmp29, tmp28, tmp27);
                        }
                    }
                    ()
                    ;
```
the line that fails to build is `auto tmp28 = tmp25.template cast<float,1>();`, because Vectorized doesn't have the `cast` method, only VecMask.

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok commented:
> Inductor seems confused about VecMasks in general; it looks like it's intermixing them with Vectorized too freely.

My instinct here is to propose that we rename all of VecMask's masking-specific methods to make it clear that they're masking-specific (possibly just converting them to free functions to be unambiguous) and then use VectorizedN instead of VecMask so that the two can't be confused accidentally. My quick triage:

- static VecMask::from() method converting to VecMask from VectorizedN<U, L>; can make into a global function that 0-extends or truncates a VectorizedN as appropriate
- VecMask::to() operation (not supported on VectorizedN, can make into a global function)
- cast() operation (again, can make into a global function)
- simple all_zero/all_masked/is_masked getters
- loadu() is a masked load, but it's not static, unlike VectorizedN's loadu. renaming and making a global function would help clarify here anyway

 The one problem seems to be `VecMask::store(bool*, count)`. It always stores as though the mask was a VectorizedN<bool, L> and requires that the count is in units of bools, whereas `VectorizedN::store(void*, count)` exists and takes a count in units of Ts. (I'm not sure how cpp.py code is supposed to reliably get this right, especially given that it's mixing up VecMask and VectorizedN currently, but if there *is* a way inductor code is supposed to never ever confuse VectorizedN and VecMask then that would solve everything.)

leslie-fang-intel commented:
Hi @swolchok, thanks for the question. `VecMask` is a convenient class which helps us doing the vectorization (SIMD) operation with other data types, you can refer to the description for more details: https://github.com/pytorch/pytorch/blob/1a8b4c65ac1ff1be1e3ca5fb04f28392e79e304a/aten/src/ATen/cpu/vec/vec_mask.h#L8-L27.

leslie-fang-intel commented:
Specific to the failures `test_comprehensive_prod_cpu_bool` you mentioned above, is it a regression caused by this PR (since I didn't see too much changes of `cpp.py` in this PR)? I can take a look, if you it can also be reproduced on X86 will be helpful to debug.

swolchok commented:
>  is it a regression caused by this PR 

yes.

swolchok commented:
#138486 fixes one of the test failures this diff covers; I think it's a latent bug that we somehow didn't catch in existing tests.

> My instinct here is to propose

This was probably overly aggressive; looks like we just have a couple paths that should have used VecMask but don't.

swolchok commented:
> if it can also be reproduced on X86

@leslie-fang-intel #138542 reproduces (but doesn't yet fix) the `test_comprehensive_prod_cpu_bool` bug on x86 on the main branch, without requiring this PR.

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use 128-bit vectors for ARM64"


The correct vector length for ARM64 is 128 bits (16
bytes). We were previously using double this, apparently just because
that would be the same length as AVX2.

Differential Revision: [D63984039](https://our.internmc.facebook.com/intern/diff/D63984039/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 malfet snadampal milpuz01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge
